src/ostree/ot-admin-builtin-finalize-staged.c

```diff
@@ -50,13 +65,56 @@ ot_admin_builtin_finalize_staged (int argc, char **argv, OstreeCommandInvocation
 
   g_autoptr(GOptionContext) context = g_option_context_new ("");
   g_autoptr(OstreeSysroot) sysroot = NULL;
+
+  /* First parse the args without loading the sysroot to see what options are
+   * set. */
   if (!ostree_admin_option_context_parse (context, options, &argc, &argv,
-                                          OSTREE_ADMIN_BUILTIN_FLAG_SUPERUSER,
+                                          OSTREE_ADMIN_BUILTIN_FLAG_NO_LOAD,
                                           invocation, &sysroot, cancellable, error))
     return FALSE;
 
-  if (!ostree_cmd__private__()->ostree_finalize_staged (sysroot, cancellable, error))
-    return FALSE;
+  if (opt_hold)
+    {
+      /* Load the sysroot unlocked so that a separate namespace isn't
+       * created. */
+      if (!ostree_admin_sysroot_load (sysroot,
+                                      OSTREE_ADMIN_BUILTIN_FLAG_SUPERUSER | OSTREE_ADMIN_BUILTIN_FLAG_UNLOCKED,
+                                      cancellable, error))
+        return FALSE;
+
+      /* In case it's an automount, open /boot so that the automount doesn't
+       * expire until before this process exits. If it did expire and got
+       * unmounted, the service would be stopped and the deployment would be
+       * finalized earlier than expected.
+       */
+      int sysroot_fd = ostree_sysroot_get_fd (sysroot);
+      glnx_autofd int boot_fd = -1;
+      g_debug ("Opening /boot directory");
+      if (!glnx_opendirat (sysroot_fd, "boot", TRUE, &boot_fd, error))
+        return FALSE;
+
+      /* We want to keep /boot open until the deployment is finalized during
+       * system shutdown, so block on SIGTERM under the assumption that it will
+       * be received when systemd stops the unit.
+       */
+      gboolean running = TRUE;
+      g_unix_signal_add (SIGTERM, sigterm_cb, &running);
+      g_print ("Waiting for SIGTERM\n");
+      while (running)
+        g_main_context_iteration (NULL, TRUE);
```

Comment:
I'm fine with this as is, but in general I think it's even cleaner where possible to *not* install a `SIGTERM` handler and let the kernel simply kill the process.  That way there's one fewer context switch at shutdown time.

IOW, we could just do:

```
while (true)
  g_main_context_iteration (NULL, TRUE);
```

Reply:
Very true. I guess I like the `SIGTERM` handler so that the service exits gracefully instead of showing as failed, although in this case you'd probably never look at the status of the service.

Reply:
What would show as failed?

```
[root@cosa-devsh ~]# cat /etc/systemd/system/testunit.service
[Service]
ExecStart=sleep infinity
[root@cosa-devsh ~]# systemctl start testunit
[root@cosa-devsh ~]# systemctl status testunit
● testunit.service
     Loaded: loaded (/etc/systemd/system/testunit.service; static)
     Active: active (running) since Tue 2022-08-30 20:21:23 UTC; 1s ago
   Main PID: 1792 (sleep)
      Tasks: 1 (limit: 1042)
     Memory: 284.0K
        CPU: 1ms
     CGroup: /system.slice/testunit.service
             └─ 1792 sleep infinity

Aug 30 20:21:23 cosa-devsh systemd[1]: Started testunit.service.
[root@cosa-devsh ~]# systemctl stop testunit
[root@cosa-devsh ~]# systemctl status testunit
○ testunit.service
     Loaded: loaded (/etc/systemd/system/testunit.service; static)
     Active: inactive (dead)

Aug 30 20:21:27 cosa-devsh systemd[1]: Stopping testunit.service...
Aug 30 20:21:27 cosa-devsh systemd[1]: testunit.service: Deactivated successfully.
Aug 30 20:21:27 cosa-devsh systemd[1]: Stopped testunit.service.
```

Reply:
PR in https://github.com/ostreedev/ostree/pull/2704

Reply:
I guess I was thinking that systemd would see the non-0 exit status and interpret it as failed. But it makes sense since it's the parent and can process the status appropriately.